File: xor/xor_test.go
package xor

import (
	"encoding/hex"
	"testing"
)

type text struct {
	plaintext  string
	ciphertext string
}

var texts = [...]text{
	{
		"c104f40c",
		"c104f406",
	},
	{
		"c10400ff",
		"c1040001",
	},
	{
		"c110fa044b15d2a7d92493a24d752580",
		"c110fa00536572766572204e65777300",
	},
	{
		"c109fa118bff558bec",
		"c109fa15865acae2c4",
	},
}

func Test(t *testing.T) {
	for _, v := range texts {
		ciphertext, err := hex.DecodeString(v.plaintext)
		if err != nil {
			t.Errorf("convert hex string failed, %s", err.Error())
		}
		Enc(ciphertext, 3, len(ciphertext)-1)
		if hex.EncodeToString(ciphertext) != v.ciphertext {
			t.Errorf("encrypt failed, expected(%s) got(%s)", v.ciphertext, hex.EncodeToString(ciphertext))
		}

		plaintext, err := hex.DecodeString(v.ciphertext)
		if err != nil {
			t.Errorf("convert hex string failed, %s", err.Error())
		}
		Dec(plaintext, 3, len(plaintext)-1)
		if hex.EncodeToString(plaintext) != v.plaintext {
			t.Errorf("encrypt failed, expected(%s) got(%s)", v.plaintext, hex.EncodeToString(plaintext))
		}
	}
}

File: wireshark/xor/xor_test.lua
local array = require("util.array")
local xor = require("xor")

local tests = {
    {
        plaintext  = "C104F40C",
        ciphertext = "C104F406",
    },
    {
        plaintext  = "C10400FF",
        ciphertext = "C1040001",
    },
    {
        plaintext =  "C110FA044B15D2A7D92493A24D752580",
        ciphertext = "C110FA00536572766572204E65777300",
    },
    {
        plaintext  = "C109FA118BFF558BEC",
        ciphertext = "C109FA15865ACAE2C4",
    }
}

for k,v in pairs(tests) do
    local ciphertext = array.fromHex(v.plaintext)
    xor.enc(ciphertext, 4, #ciphertext)
    assert(v.ciphertext == array.toHex(ciphertext),
    string.format("xor failed! expected(%s) got(%s)", v.ciphertext, array.toHex(ciphertext)))

    local plaintext = array.fromHex(v.ciphertext)
    xor.dec(plaintext, 4, #plaintext)
    assert(v.plaintext == array.toHex(plaintext),
    string.format("xor failed! expected(%s) got(%s)", v.plaintext, array.toHex(plaintext)))
end